### ISSUE TITLE
Add extra API for offline/online phases in ECDSA signing

### ISSUE THREAD

**Issue by RandomLattice** [bitcoin-core/secp256k1#1610]
We'd love to split the ECDSA signing operation into two steps:
- an "offline" step that is independent on the message,
- an "online" step that depends on the message.

The advantage here is that the offline step can be precomputed arbitrarily early and do "most of the computational work". This is very useful for signers that have a lot of idle time before the signing request comes in, but little time to compute and return the signature. For example, hardware wallets.

In ECDSA, the offline step can essentially compute nonce generation + scalar multiplication + nonce modular inversion. These values ("precomputed material") are input to the online part. The online part just finishes the ECDSA computation. The online part is extremely fast (we're seeing around 5000x faster than the offline phase in a non-libsecp256k1 proof-of-concept).

The precomputed material is secret, can only be used once and needs to be wiped after usage.

API-wise this would probably mean adding two functions: 
 * `secp256k1_ecdsa_sign_split_phase_precompute()`
 * `secp256k1_ecdsa_sign_split_phase_online()`

The composition of the two functions compute `secp256k1_ecdsa_sign()`. The input/output behavior does not change. 

Is there any appetite for this? Happy to write the code for this.

**Comment by real-or-random** [comment]
> The precomputed material is secret, can only be used once and needs to be wiped after usage.

Indeed, and this is the drawback of this. It's a major footgun, and I'm sure this is the reason why it's currently implemented this way.

Personally, I'm a bit on the fence. We're about to add a [MuSig2 module](https://github.com/bitcoin-core/secp256k1/pull/1479), which is also prone to nonce-reuse failures -- but in that case, it is unavoidable to split nonce generation and the actual signing into two separate steps. In the end, I don't think I'm against giving users advanced and dangerous features, but they should be clearly marked as such.

I'd love to hear your use case and scenario here. Noone has asked for this before, and I doubt many users will benefit. (And anyway, ECDSA can be considered somewhat deprecated in Bitcoin, but okay, the same could be done for Schnorr signing). Do you personally need/want this, or is this is just a suggestion for improvement?

It would be interesting to know what other contributors and maintainers think.

**Comment by jmecom** [comment]
> @jmecom Sorry for the slight tangent here, but:
> 
> > On our chip, ECDSA takes ~60ms and Schnorr takes ~55ms. We can't use large precomputation tables due to limited RAM.
> 
> Since #1058 landed in 0.5.0, the supported table sizes (for the key generation/signing code) are 2 KiB, 22 KiB, and 86 KiB. How much is acceptable to you? It's easy to add more configurations if there is a need for them.

I just pulled in 0.5.1 and experimented with this. I didn't see a huge impact on the table size with a window size of two:
```
    62 ms @ 2kb
    58 ms @ 22kb
	56 ms @ 86kb
```

Window size of five took about 50ms @ 22kb. So, I don't think more fine-grained options are really necessary for our usecase at least.

**Comment by jmecom** [comment]
> @jmecom By "window size", you mean the `ECMULT_WINDOW_SIZE` setting? That should have no impact on key generation/signing speed, only on verification speed.

Ah, woops. Due to the way I'm profiling, there is a little variation across each run, so I mistakenly thought that lowered the duration, but it was just an anomaly. Ignore that, but the rest is accurate.

**Comment by apoelstra** [comment]
The above proposal sounds reasonable to me (other than the name -- I agree "precomputednonce" or "detachednonce" sound better).

But it does open a question about whether we would want an "off by default" module. Currently we have two categories of modules:

* "experimental" modules where we do not guarantee API stability, but hope to eventually get there
* non-experimental modules, which are all turned on for releases, since we don't have a good semver-compatible way for distributions to pick and choose which modules they compile in

Originally I think we had imagined that users would be able to pick and choose what modules they need, but that's not really compatible with shared library distribution; and we also hoped that users would be able to write their own custom modules, but that doesn't really work with C tooling (though I suppose it could be made to work with nix stdenv and the `patches` field maybe..).

Anyway, this module would presumably be "experimental", but unlike the other experimental modules, I don't think we would ever be willing to promote it to a "real" module.
